### ISSUE TITLE
Allow custom patterns for `bun test`

### ISSUE THREAD

**Issue by swissspidy** [oven-sh/bun#3440]
### What is the problem this feature would solve?

Right now Bun looks only for files with these patterns:

* `*.test.{js|jsx|ts|tsx}`
* `*_test.{js|jsx|ts|tsx}`
* `*.spec.{js|jsx|ts|tsx}`
* `*_spec.{js|jsx|ts|tsx}`

It would be nice if this list could be customized somehow. For example, I usually just put all files in a `test/` or `specs/` directory and the files themselves don't have any specific suffix.

### What is the feature you are proposing to solve the problem?

Allow customizing the list of patterns, for example in the config file or via CLI args. Including glob support.

### What alternatives have you considered?

If #1508 lands, I could manually grab a list of files and pass them all to the test command, but that's a bit cumbersome.

**Comment by simon-abbott** [comment]
Agreed. Our codebase separates our tests into `__tests__/*.unit.ts` and `__tests__/*.int.ts` files. Jest and Vitest both allow us to specify a custom glob to match those. This is a really important piece to allow for near-seamless transitions over to Bun. Auto-rewriting the `vitest` imports doesn't reduce friction if I have to rename all the files anyway. 🙃

**Comment by xavikortes** [comment]
+1. Our codebase is divided into different modules and into `*.test.ts` and `*.itest.ts`. We have to rename `*.itest.ts` to `*.spec.ts`. Also can't filter by both of those patterns at the same time.

**Comment by Jarred-Sumner** [comment]
We can do this now that @zackradisic implemented support for globs

**Comment by paperclover** [comment]
i want to note as a workaround, you can pass exact paths to the test runner and it will ignore the pattern. this was introduced in i think 1.0.16 but i might be mistaken on when.

this in combination with **built-in shell glob expansion** can allow custom test patterns:

(note all the files i have are empty)
<img width="464" alt="image" src="https://github.com/oven-sh/bun/assets/24465214/34bd7f26-d8c8-49d4-8a12-8b62001e92a1">

i dont think this is good enough the close the issue, it should be configurable so that plain `bun test` works

**Comment by anru** [comment]
+1 on this. We have both playwright and jest tests in same directories and divide them only via testMatch patterns.

**Comment by xavikortes** [comment]
Is there any workaround to run tests that match BOTH "directory filter" AND "test type"?
(e.g.: bun test ./test/app .spec.ts)
Right now it is running all that match EITHER one OR another. Is there any way to do it without glob?

**Comment by MarceloPrado** [comment]
+1, this would be great to allow separating unit tests and integration tests. My unit tests are `*.test.ts` and integration `*.integration.test.ts`

**Comment by ivan-kleshnin** [comment]
How to exclude a folder with `bun test`? There's no config file 🤯

**Comment by s-h-a-d-o-w** [comment]
> this in combination with **built-in shell glob expansion** can allow custom test patterns:
> 
> (note all the files i have are empty) <img alt="image" width="464" src="https://private-user-images.githubusercontent.com/24465214/290496678-34bd7f26-d8c8-49d4-8a12-8b62001e92a1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc3MDY5OTAsIm5iZiI6MTcxNzcwNjY5MCwicGF0aCI6Ii8yNDQ2NTIxNC8yOTA0OTY2NzgtMzRiZDdmMjYtZDhjOC00OWQ0LThhMTItOGI2MjAwMWU5MmExLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjA2VDIwNDQ1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ2ZWQ5NWJmZGRkOTc5N2U5NTM1YTEwNGFkOTc3ZDdkYmYxNDdmOWYyNDAxNThlNmJjM2Q4ODBhOWQyMzhmNTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.x7tmzb-NZ3oxwVptqiqSbMVO9QooBircL9m4oKq_EDA">

Looks like this has changed since:
```
$ bun test ./**/*.test.ts
bun test v1.1.12 (43f0913c)
error: ModuleNotFound resolving "./**/*.test.ts" (entry point)

 0 pass
 0 fail
Ran 0 tests across 0 files. [7.00ms]
```

**Comment by spacedawwwg** [comment]
> > this in combination with **built-in shell glob expansion** can allow custom test patterns:
> > (note all the files i have are empty) <img alt="image" width="464" src="https://private-user-images.githubusercontent.com/24465214/290496678-34bd7f26-d8c8-49d4-8a12-8b62001e92a1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc3MDY5OTAsIm5iZiI6MTcxNzcwNjY5MCwicGF0aCI6Ii8yNDQ2NTIxNC8yOTA0OTY2NzgtMzRiZDdmMjYtZDhjOC00OWQ0LThhMTItOGI2MjAwMWU5MmExLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjA2VDIwNDQ1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ2ZWQ5NWJmZGRkOTc5N2U5NTM1YTEwNGFkOTc3ZDdkYmYxNDdmOWYyNDAxNThlNmJjM2Q4ODBhOWQyMzhmNTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.x7tmzb-NZ3oxwVptqiqSbMVO9QooBircL9m4oKq_EDA">
> 
> Looks like this has changed since:
> 
> ```
> $ bun test ./**/*.test.ts
> bun test v1.1.12 (43f0913c)
> error: ModuleNotFound resolving "./**/*.test.ts" (entry point)
> 
>  0 pass
>  0 fail
> Ran 0 tests across 0 files. [7.00ms]
> ```

Experiencing the exact same issue

`"test:unit": "bun test ./**/*.unit.test.ts"`

```SHELL
$ bun test ./**/*.unit.test.ts
bun test v1.1.29 (6d43b366)
error: ModuleNotFound resolving "./**/*.unit.test.ts" (entry point)

 0 pass
 0 fail
Ran 0 tests across 0 files. [4.00ms]
```

**Comment by kevinclarkadstech** [comment]
Yes, what happened? 😢 I am also getting this when trying to do the exact as above to filter unit vs. integration tests.

**Comment by kevinclarkadstech** [comment]
Nevermind, I am able to get it working!!  

```
 "test": "bun test:unit && bun test:integration",
    "test:unit": "bun test ./**/*.unit.test.*",
```

Is doing the proper filtering for me.  Thank you everyone who is building Bun, it's a great product! 👏 

P.S. I did find another cool thing, if you would for whatever reason prefer to filter tests by describe() instead of filename.  I.e. you could say

`describe('verifyAuthToken.unit', () => {});`

and say "bun test .unit" and it will filter all tests described with .unit in the name.  👍

**Comment by sepirak-nick** [comment]
This isn't working for me, when running through a package.json script specifically:

```JSON
"scripts": {
  "test:e2e": "bun test ./**/*.e2e-spec.*"
}
```

```shell
> bun test:e2e
bun test v1.2.0 (b0c5a765)
error: ModuleNotFound resolving "./**/*.e2e-spec.*" (entry point)
error: script "test:e2e" exited with code 1
```

**Comment by Necrelox** [comment]
Hello, has there been any feedback on this problem?

**Comment by Necrelox** [comment]
![Image](https://github.com/user-attachments/assets/ec98d9d3-4460-461f-9799-54618016111f)
 A workaround : 
```json
    "test:unit": "bun test --coverage $(fd .spec.ts test/unit)",
// or
    "test:unit": "bun test --coverage $(find test/unit -name '*.spec.ts')",
```

**Comment by RyanGst** [comment]
>     "test:unit": "bun test --coverage $(fd .spec.ts test/unit)",
> // or
>     "test:unit": "bun test --coverage $(find test/unit -name '*.spec.ts')",

thanks @Necrelox, this works fine for us

**Comment by joakimbeng** [comment]
I need this as well!

---

In my case I'm working on cucumber support that uses `bun:test` as the test runner, instead of a custom test runner that `cucumber-js` uses, so that I for instance can utilise Bun's snapshot testing and other nice features.

I have written a Bun Plugin that's able to load `*.feature` files that generates test files using `describe`, `it`, etc. from `bun:test`, but the problem is that Bun's hardcoded glob pattern for test files makes it impossible to run these files directly.

My current workaround is to have an empty test file "features.test.ts" containing:

```ts
const featuresGlob = new Glob('**/*.feature');

for await (const file of featuresGlob.scan({ absolute: true })) {
  await import(file);
}
```

This works and the feature tests are running as expected, but all test snapshots are put in the same `features.test.ts.snap` file , which doesn't scale well, no matter what feature file they came from.
Also the watch mode is not working, changing a `*.feature` file does not re-run the tests, for that I have to re-save the `features.test.ts` file.

If i try to run a feature file directly by passing its path to `bun test` it complains about the file not matching a specific pattern (even though it actually does as it has "_test_" as part of the file name):

```
bun test v1.2.23-canary.86 (f45900d7)
The following filters did not match any test files:
 features/my_test_first.feature
20 files were searched [67.00ms]

note: Tests need ".test", "_test_", ".spec" or "_spec_" in the filename (ex: "MyApp.test.ts")
```

**To conclude:** the default glob pattern for test files should be customisable _and_ if there's a registered loader that can handle a specific file extension Bun shouldn't assume it can only handle JS/TS files, when passing a file path directly to the cli for instance.

Take Vitest as a comparison where you can configure the `test.include` option to something like `['**/*.feature', '**/*.test.ts']`, then have a transform plugin that handles the `*.feature` files and transforms them to JavaScript, and it'll find and run the feature files without any problems or extra cli arguments.

**Comment by janhesters** [comment]
I created a PR for this: https://github.com/oven-sh/bun/pull/26714
